spec/models/user_spec.rb

```diff
@@ -36,6 +36,18 @@
     end
   end
 
+  describe 'auto_approve' do
+    describe 'known_email_domain' do
+      it do
```

Comment:
Faire deux `it`, un pour chaque cas, et donner un objet à chaque `it` (ou utiliser un subject). Je ferais comme ça :

```ruby
# Avec un subject
let(:user) { create(:user, :just_registered, email: email) }
subject { user.is_approved? }

context "with an unkown email domain" do
  let(:email) { "nico@bou.io" }

  it { expect(subject).to be_falsey }
end

context "with a kown email domain" do
  let(:email) { "nico@beta.gouv.fr" }

  it { expect(subject).to be_truthy }
end

# Sans subject
let(:user) { create(:user, :just_registered, email: email) }

context "with an unknown email domain" do
  let(:email) { "nico@bou.io" }

  it "does not automatically approve the user" do
    expect(user.is_approved?).to be_falsey
  end
end

context "with a known email domain" do
  let(:email) { "nico@beta.gouv.fr" }

  it "automatically approves the user" do
    expect(user.is_approved?).to be_truthy
  end
end
```

Ensuite se pose la question de ce qu'on teste. Là quelque part on teste Rails : on teste que quand on fait un create, ça appelle bien la méthode qu'on a défini en callback ;  et on teste ensuite le comportement de cette méthode. Je suis assez partagé. D'un côté je me dis "c'est pas grave, ça fait pas de mal, et au moins ça vérifie que j'ai bien spécifié le callback" (un genre de test unitaire++ quoi). De l'autre, c'est vrai qu'on teste inutilement une partie de Rails, et qu'on pourrait ne tester que `#mon_callback`. Donc perso je pense que je partirais sur (dans le cas "on utilise subject") :
```ruby
describe "#auto_approve_if_known_domain!" do
  let(:user) { build(:user, :just_registered, email: email) }
  subject { user.is_approved? }

  before do
    user.auto_approve_if_known_domain!
  end

  context "with an unkown email domain" do
    let(:email) { "nico@bou.io" }

    it { expect(subject).to be_falsey }
  end

  context "with a kown email domain" do
    let(:email) { "nico@beta.gouv.fr" }

    it { expect(subject).to be_truthy }
  end
end
```

Reply:
Sur le formalisme des tests, le projet était resté au milieu du gué, j’ai l’impression. J’ai gardé la façon de faire de `user_spec.rb` mais d’autres fichiers sont plus structurés.

(Si on va plus loin, on devrait tester le fait qu’un adresse est whitelistée ou non, indépendamment de User.)

Reply:
> Si on va plus loin, on devrait tester le fait qu’un adresse est whitelistée ou non, indépendamment de User.

C'est à dire ?

Reply:
L’algorithme (c’est un bien grand mot) qu’on teste au fond, c’est « Est-ce que cette adresse est whitelistée? ». On peut très bien faire ça sans User.

Reply:
Oui, mais il faudra dans tous les cas tester son application au cas de User je pense